<?xml version="1.0" encoding="UTF-8"?>

<launch>

    <arg name="bag" default="true"/>
    <arg name = "bagfile" default="almostButVilsonTooFast.bag"/>

    <param name="/use_sim_time" value="$(arg bag)"/>
    <node pkg="rosbag" type="play" name="rosbag_record_diag" args="--clock --loop -s 0 -r 1.0 $(find atypical_driving)/worlds/$(arg bagfile)" if = "$(arg bag)">
<!--        <remap from = "/current_pose_cov" to = "/current_pose"/>-->
    </node>

  <arg name = "param_file" default = "yugokri.yaml"/>

  <arg name = "goal/x"  default = "323891.02"/> <!--in world frame -->
  <arg name = "goal/y" default = "4145172.15"/> <!--in world frame -->

    <!-- Arguments Start -->
  <arg name = "pcl_topic" default = "/atypical_planning_test/pcl_filtered"/>
  <arg name ="snu_frame_id" default="SNU"/>
  <arg name ="occu_frame_id" default="SNU"/>
  <arg name ="world_frame_id" default="map"/>
  <arg name ="car_base_frame_id" default="base_link"/>
  <arg name ="lidar_frame_id" default = "velodyne"/>
  <arg name ="detected_object_frame_id" default = "velodyne"/>

  <arg name = "lane_width" default="4.5"/>
  <arg name = "lane_csv_file" default = "$(find atypical_driving)/lane/interpolated_yugokri_path1.csv"/>
  <arg name = "log_file_prefix" default= "$(find atypical_driving)/log/log" />


  <arg name = "period" default= "0.1"/>
  <arg name = "horizon"  default= "5"/> <!--Also change dimension.h-->
  <arg name="planning_dt" default="0.1"/>

  <group ns="atypical_planning_test">
    <rosparam file="$(find atypical_driving)/param/MPC_weight.yaml"/>
  </group>


    <!-- run pointcloud_to_laserscan node -->
    <node pkg="pointcloud_to_laserscan" type="pointcloud_to_laserscan_node" name="pointcloud_to_laserscan">
        <remap from="cloud_in" to="$(arg pcl_topic)"/>
        <remap from="scan" to="/scan_snu"/>
         <rosparam command="load" file="$(find atypical_driving)/param/$(arg param_file)"/>
    </node>

    <!--  1. OccupancyGrid  -->
     <include file = "$(find atypical_driving)/launch/occupancy2d/costmap_2d.launch">
         <arg name = "play_bag" value = "false"/>
         <arg name = "scan_topic" value = "scan_snu"/>
         <arg name = "use_sim_time" value="false"/>
         <arg name = "rviz" value = "false"/>
         <arg name = "occupancy_frame_id" value = "$(arg occu_frame_id)"/>
         <arg name = "lidar_frame_id" value = "$(arg lidar_frame_id)"/>
     </include>


    <!--2. snu_car_tf to velodyne_tf -->
    <node pkg="tf" type="static_transform_publisher" name="link2lidar" args="-1.0.0 0 0.2 0 0 0  $(arg car_base_frame_id) $(arg lidar_frame_id) 100"/>

    <!--3. client-->
    <node pkg="atypical_driving" type="keti_client" name="atypical_driving_client" output="screen">
        <param name="rejection_radius" value="1.6"/>
    </node>

    <!--3. server-->
    <node pkg="atypical_driving" type="integration_server" name="atypical_planning_test" output="screen">

        <remap from = "/occupancy_grid" to = "/costmap_node/costmap/costmap"/>
        <remap from = "/current_pose" to = "/current_pose_cov"/>
        <remap from = "/vehicle_cmd" to = "/vehicle_cmd_snu"/>

        <param name = "world_frame_id" value="$(arg world_frame_id)" />
        <param name = "snu_frame_id" value="$(arg snu_frame_id)" />
        <param name = "occu_map_frame_id" value="$(arg occu_frame_id)" />
        <param name = "base_link_id" value="$(arg car_base_frame_id)" />
        <param name = "detected_objects_id" value="$(arg detected_object_frame_id)" />

        <param name = "local_planner/obstacle_radius_nominal" value="1"/>

        <!--Shaping params for nominal_speed selection-->
        <param name = "vmax" value = "3.0"/> <!--m/s-->
        <param name = "vmin" value = "2.0"/> <!--m/s-->
        <param name = "v_ref_past_weight" value = "0.8"/> <!--m/s-->
        <param name = "curve_thres" value = "0.05"/> <!--m/s-->

        <param name = "goal/x" value = "$(arg goal/x)"/>
        <param name = "goal/y" value = "$(arg goal/y)"/>
        <param name = "goal_thres" value = "1.5"/>

        <param name = "lane_csv_file" value = "$(arg lane_csv_file)"/>
        <param name = "lane_width" value="$(arg lane_width)"/>
        <param name = "log_file_prefix" value = "$(arg log_file_prefix)" />

        <param name="global_planner/car_width" value="1.0" />
        <param name="global_planner/period" value="$(arg period)" />
        <param name="global_planner/grid_resolution" value="0.3" />
        <param name="global_planner/max_steering_angle" value = "$(eval 3.14/180*9)"/>
        <param name="global_planner/smoothing_distance" value="8.0" />
        <param name="global_planner/smoothing_cliff_min_bias" value="0.5" />
        <param name="global_planner/smoothing_cliff_ratio" value="0.5"/>
        <param name="global_planner/start_smoothing_distance" value="6.0" />
        <param name="global_planner/corridor_width_min" value = "0.5"/>
        <param name="global_planner/corridor_width_dynamic_min" value = "8.0"/>
        <param name="global_planner/corridor_max_bias" value = "100"/>
        <param name="global_planner/safe_distance" value = "6.0"/>
        <param name="global_planner/nominal_acceleration" value = "0.5"/>
        <param name="global_planner/object_velocity_threshold" value="0.1"/>
        <param name="global_planner/max_obstacle_prediction_query_size" value="1000"/>
        <param name="global_planner/acc_stop_distance" value="3.0"/>
        <param name="global_planner/acc_stop_angle" value="$(eval 3.14/180*5)"/>

        <param name="local_planner/horizon" value="$(arg horizon)" />
        <param name="local_planner/ts" value="$(arg planning_dt)" />
        <param name="local_planner/max_steer" value="0.52" />
        <param name="local_planner/max_accel" value="1" />
        <param name="local_planner/min_accel" value="-3" />
        <param name="local_planner/N_corr" value="5" />
        <param name="local_planner/isRearWheel" value="false"/>
        <param name="local_planner/dyn_obst_range" value="30.0"/>
    	<param name="local_planner/sfc_margin" value="0.5"/>
    	<param name="local_planner/same_wpts_distance" value="0.01"/>
        <param name="local_planner/sameNumPointStopping" value="70"/>
	    <!--smoothing_type: 0:Exponential Average(0924test), 1:Moving Average, 2:Acc(exp),Steer(mov), 3:ignore small handling -->
	    <!--ignore_angle, 1deg: 0.017, 2deg: 0.035, 3deg: 0.052, 4deg: 0.070 -->
	    <param name = "smooth_weight" value="0.7"/>
        <param name = "moving_horizon" value="20"/>
        <param name = "smoothing_type" value="2"/>
	    <param name = "ignore_angle" value="0.017"/>
        <param name = "stopSpeed" value="0.5"/>


        <param name="predictor/observation_queue" value="10" />
        <param name="predictor/ref_height" value="0.4" />
        <param name="predictor/log_dir" value="$(find atypical_driving)/log/predictor" />

        <rosparam command="load" file="$(find atypical_driving)/param/$(arg param_file)"/>

    </node>

  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find atypical_driving)/launch/rviz_config/config.rviz" output="log" />


</launch>
